In [623]:
from google.colab import drive

drive.mount("/content/gdrive")  

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [624]:
%cd "/content/gdrive/MyDrive/IA/Estadistica"
%ls

/content/gdrive/MyDrive/IA/Estadistica
bcase.html  bcase.ipynb  Datos.csv  Entregable2.ipynb  Firmas.csv


In [625]:
import pandas as pd
import numpy as np
import seaborn as sns


In [626]:
df_firms = pd.read_csv('Firmas.csv')
df_data = pd.read_csv('Datos.csv')


In [627]:
df_firms.head()

Ticker                   Name  N  Class Country\nof Origin Type of Asset  \
0   FLWS  1 800 Flowers Com Inc  1  Com A                 US         Stock   
1    TXG      10x Genomics, Inc  2  Com A                 US         Stock   
2   GOED      1847 Goedeker Inc  3    Com                 US         Stock   
3   ONEM  1life Healthcare, Inc  4    Com                 US         Stock   
4   SRCE        1st Source Corp  5    Com                 US         Stock   

               Sector NAICS\nlevel 1 Exchange / Src    Sector\nEconomatica  \
0                       Retail Trade         NASDAQ                  Trade   
1                      Manufacturing         NASDAQ      Electric Electron   
2                       Retail Trade           AMEX                  Trade   
3  Health Care and Social Assistance         NASDAQ                  Other   
4              Finance and Insurance         NASDAQ  Finance and Insurance   

                        Sector NAICS\nlast available partind  
0                                       Retail Trade       -  
1  Navigational, Measuring, Electromedical, and C...       -  
2              Furniture and Home Furnishings Stores       -  
3                              Offices of Physicians       -  
4                   Depository Credit Intermediation       -

In [628]:
df_data.head()

firm       q    revenue       cogs       sgae  otheropexp  extraincome  \
0    A  2000q1        NaN        NaN        NaN         NaN          NaN   
1    A  2000q2  2485000.0  1261000.0  1010000.0         0.0      42000.0   
2    A  2000q3  2670000.0  1369000.0  1091000.0         0.0      28000.0   
3    A  2000q4  3372000.0  1732000.0  1182000.0         0.0      10000.0   
4    A  2001q1  2841000.0  1449000.0  1113000.0         0.0      -6000.0   

   finexp  incometax  totalassets  totalliabilities  shortdebt  longdebt  \
0     NaN        NaN          NaN               NaN        NaN       NaN   
1     0.0    90000.0    7321000.0         2679000.0   512000.0       0.0   
2     0.0    83000.0    7827000.0         2925000.0   528000.0       0.0   
3     0.0   163000.0    8425000.0         3160000.0   830000.0       0.0   
4     0.0   119000.0    9208000.0         3667000.0   556000.0       0.0   

   stockholderequity   adjprice  originalprice  sharesoutstanding  \
0                NaN  63.761670       104.0000         452000.000   
1          4642000.0  45.215607        73.7500         452271.967   
2          4902000.0  30.003238        48.9375         453014.579   
3          5265000.0  33.566841        54.7500         456366.381   
4          5541000.0  18.840347        30.7300         456769.737   

   fiscalmonth  year  cto  
0          NaN  2000    1  
1          6.0  2000    2  
2          9.0  2000    3  
3         12.0  2000    4  
4          3.0  2001    1

In [629]:
df_firms.columns = df_firms.columns.str.replace('Ticker', 'firm')


In [630]:
df_mrg = pd.merge(df_firms, df_data, on="firm")
df_mrg

firm                   Name     N  Class Country\nof Origin  \
0       FLWS  1 800 Flowers Com Inc     1  Com A                 US   
1       FLWS  1 800 Flowers Com Inc     1  Com A                 US   
2       FLWS  1 800 Flowers Com Inc     1  Com A                 US   
3       FLWS  1 800 Flowers Com Inc     1  Com A                 US   
4       FLWS  1 800 Flowers Com Inc     1  Com A                 US   
...      ...                    ...   ...    ...                ...   
323811  ZNGA              Zynga Inc  3608  Com A                 US   
323812  ZNGA              Zynga Inc  3608  Com A                 US   
323813  ZNGA              Zynga Inc  3608  Com A                 US   
323814  ZNGA              Zynga Inc  3608  Com A                 US   
323815  ZNGA              Zynga Inc  3608  Com A                 US   

       Type of Asset Sector NAICS\nlevel 1 Exchange / Src Sector\nEconomatica  \
0              Stock          Retail Trade         NASDAQ               Trade   
1              Stock          Retail Trade         NASDAQ               Trade   
2              Stock          Retail Trade         NASDAQ               Trade   
3              Stock          Retail Trade         NASDAQ               Trade   
4              Stock          Retail Trade         NASDAQ               Trade   
...              ...                   ...            ...                 ...   
323811         Stock           Information         NASDAQ     Software & Data   
323812         Stock           Information         NASDAQ     Software & Data   
323813         Stock           Information         NASDAQ     Software & Data   
323814         Stock           Information         NASDAQ     Software & Data   
323815         Stock           Information         NASDAQ     Software & Data   

                          Sector NAICS\nlast available  ... totalliabilities  \
0                                         Retail Trade  ...          76657.0   
1                                         Retail Trade  ...          65723.0   
2                                         Retail Trade  ...          75163.0   
3                                         Retail Trade  ...          88231.0   
4                                         Retail Trade  ...          83160.0   
...                                                ...  ...              ...   
323811  Data Processing, Hosting, and Related Services  ...        3111600.0   
323812  Data Processing, Hosting, and Related Services  ...        3060700.0   
323813  Data Processing, Hosting, and Related Services  ...        3247000.0   
323814  Data Processing, Hosting, and Related Services  ...        2953900.0   
323815  Data Processing, Hosting, and Related Services  ...              NaN   

       shortdebt   longdebt  stockholderequity   adjprice  originalprice  \
0         4380.0     8869.0           171804.0   7.125000       7.125000   
1         1839.0     9441.0           158918.0   5.125000       5.125000   
2         7654.0    10916.0           136286.0   4.984375       4.984375   
3         2157.0    10452.0           125752.0   4.125000       4.125000   
4         4777.0    12318.0           117240.0   8.000000       8.000000   
...          ...        ...                ...        ...            ...   
323811   17700.0  1457000.0          3020600.0  10.630000      10.630000   
323812   17600.0  1467600.0          3007000.0   7.530000       7.530000   
323813   17100.0  1477200.0          3111900.0   6.400000       6.400000   
323814   17900.0  1673700.0          2906100.0   9.240000       9.240000   
323815       NaN        NaN                NaN   8.180000       8.180000   

        sharesoutstanding  fiscalmonth  year  cto  
0               64097.913          9.0  2000    1  
1               64170.913         12.0  2000    2  
2               64185.913          3.0  2000    3  
3               64187.163          6.0  2000    4  
4               64187.163          9.0  2001    

In [631]:
df_mrg = df_mrg.drop(['N', 'Class', 'Country\nof Origin', "Type of Asset", "partind", "Sector NAICS\nlast available", "Exchange / Src", "year", "cto"], axis=1)


In [632]:
df_mrg["Book value"] = df_mrg["totalassets"] - df_mrg["totalliabilities"]
df_mrg["Market value"] = df_mrg["originalprice"] * df_mrg["sharesoutstanding"]

**Operating profit margin**

In [633]:
df_mrg["Operating profit margin"] = df_mrg["revenue"] - df_mrg["cogs"] - df_mrg["sgae"]

**Ebit**

In [634]:
df_mrg["Ebit"] = df_mrg["revenue"] - df_mrg["cogs"] - df_mrg["sgae"] - df_mrg["otheropexp"]

**OPM**

In [635]:
df_mrg["OPM"] = df_mrg["Ebit"] / df_mrg["revenue"] 

**Net income**

In [636]:
df_mrg["Net income"] = df_mrg["Ebit"] - df_mrg["incometax"] - df_mrg["finexp"]

**Profit margin**

In [637]:
df_mrg["Profit margin"] = df_mrg["Net income"] / df_mrg["revenue"] 

In [638]:
df_mrg['qdate'] = pd.PeriodIndex(df_mrg["q"], freq="Q")

df_mrg.set_index(['firm', 'qdate'], drop=False)

firm                   Name Sector NAICS\nlevel 1  \
firm qdate                                                       
FLWS 2000Q1  FLWS  1 800 Flowers Com Inc          Retail Trade   
     2000Q2  FLWS  1 800 Flowers Com Inc          Retail Trade   
     2000Q3  FLWS  1 800 Flowers Com Inc          Retail Trade   
     2000Q4  FLWS  1 800 Flowers Com Inc          Retail Trade   
     2001Q1  FLWS  1 800 Flowers Com Inc          Retail Trade   
...           ...                    ...                   ...   
ZNGA 2021Q2  ZNGA              Zynga Inc           Information   
     2021Q3  ZNGA              Zynga Inc           Information   
     2021Q4  ZNGA              Zynga Inc           Information   
     2022Q1  ZNGA              Zynga Inc           Information   
     2022Q2  ZNGA              Zynga Inc           Information   

            Sector\nEconomatica       q   revenue      cogs      sgae  \
firm qdate                                                              
FLWS 2000Q1               Trade  2000q1   85045.0   54143.0   52146.0   
     2000Q2               Trade  2000q2  123288.0   75607.0   63201.0   
     2000Q3               Trade  2000q3   72516.0   45091.0   51590.0   
     2000Q4               Trade  2000q4  134243.0   79099.0   67206.0   
     2001Q1               Trade  2001q1  103221.0   64020.0   48856.0   
...                         ...     ...       ...       ...       ...   
ZNGA 2021Q2     Software & Data  2021q2  720000.0  253300.0  391900.0   
     2021Q3     Software & Data  2021q3  704700.0  240800.0  409400.0   
     2021Q4     Software & Data  2021q4  695500.0  260100.0  436600.0   
     2022Q1     Software & Data  2022q1  691200.0  252200.0  424000.0   
     2022Q2     Software & Data  2022q2       NaN       NaN       NaN   

             otheropexp  extraincome  ...  sharesoutstanding  fiscalmonth  \
firm qdate                            ...                                   
FLWS 2000Q1         0.0        111.0  ...          64097.913          9.0   
     2000Q2         0.0         27.0  ...          64170.913         12.0   
     2000Q3         0.0         88.0  ...          64185.913          3.0   
     2000Q4         0.0        335.0  ...          64187.163          6.0   
     2001Q1         0.0         76.0  ...          64187.163          9.0   
...                 ...          ...  ...                ...          ...   
ZNGA 2021Q2         0.0      -9100.0  ...        1088762.805          6.0   
     2021Q3     66800.0       -600.0  ...        1091890.649          9.0   
     2021Q4         0.0     -10400.0  ...        1120181.490         12.0   
     2022Q1         0.0     -15700.0  ...        1132408.337          3.0   
     2022Q2         NaN          NaN  ...        1137273.677          NaN   

             Book value  Market value  Operating profit margin     Ebit  \
firm qdate                                                                
FLWS 2000Q1    171804.0  4.566976e+05                 -21244.0 -21244.0   
     2000Q2    158918.0  3.288759e+05                 -15520.0 -15520.0   
     2000Q3    136286.0  3.199267e+05                 -24165.0 -24165.0   
     2000Q4    125752.0  2.647720e+05                 -12062.0 -12062.0   
     2001Q1    117240.0  5.134973e+05                  -9655.0  -9655.0   
...                 ...           ...                      ...      ...   
ZNGA 2021Q2   3020600.0  1.157355e+07                  74800.0  74800.0   
     2021Q3   3007000.0  8.221937e+06                  54500.0 -12300.0   
     2021Q4   3111900.0  7.169162e+06                  -1200.0  -1200.0   
     2022Q1   2906100.0  1.046345e+07                  15000.0  15000.0   
     2022Q2         NaN  9.302899e+06                      NaN      NaN   

                  OPM  Net income  Profit margin   qdate  
firm qdate                                                
FLWS 2000Q1 -0.249797    -19376.0      -0.227832  2000Q1  
     2000Q2 -0.125884    -12937.0      -0.104933 

In [639]:
df_reg = df_mrg.copy()

In [640]:
df_reg["EPS"] = df_reg['Net income'] / df_reg["sharesoutstanding"]

In [641]:
df_reg["EPSP"] = (df_reg['EPS'] / df_reg["originalprice"])

In [642]:
df_reg["Sales anual growht"] = (df_reg['revenue'] / df_reg.groupby(["firm"])["revenue"].shift(4)) - 1

In [643]:
df_reg["Operating profit growht"] = (df_reg['Ebit'] / df_reg.groupby(["firm"])["Ebit"].shift(4)) - 1

In [644]:
df_reg["Operating profit margin"] = df_reg['Ebit'] / df_reg['revenue']

In [645]:
df_reg['Book to market ratio'] = df_reg['Book value'] / df_reg['Market value']

In [646]:
df_reg['Short financial leverage'] = df_reg['shortdebt'] / df_reg['totalassets']

In [647]:
df_reg['Long financial leverage'] = df_reg['longdebt'] / df_reg['totalassets']

In [648]:
df_reg["Stock annual return"] = np.log(df_reg['adjprice']) - np.log(df_reg.groupby(["firm"])['adjprice'].shift(4))

In [649]:
df_reg["F1r"] = df_reg.groupby(["firm"])['Stock annual return'].shift(-1)

In [650]:
df_reg = df_reg.replace([np.nan], 0) 

In [651]:
df_reg.replace([np.inf, -np.inf], np.nan, inplace=True) 

CREACIÓN DE LAS VARIABLES DE SIZE



In [652]:
def dense_inclusive_pct(x):
    r = x.rank(method='dense') - 1
    return r / r.max() * 100

df_reg["pct"]=df_reg.groupby('q')['Market value'].apply(dense_inclusive_pct).astype(int)


df_reg["isSmall"] = df_reg.pct <= 33
df_reg["isSmall"] = df_reg["isSmall"].astype(int)

df_reg["isMedium"] = (df_reg.pct <= 66) & (df_reg.pct > 33) 
df_reg["isMedium"] = df_reg["isMedium"].astype(int)




FILTRAR DF PARA SOLO TENER LAS EMPRESAS DE MANUFACTURA



In [653]:
is_man=df_reg[df_reg["Sector NAICS\nlevel 1"]!="Manufacturing"].index
df_reg=df_reg.drop(is_man)


In [654]:
df_regL = df_reg.loc[:, ["firm","q","Stock annual return", "F1r","EPSP", "Sales anual growht", "Operating profit growht", "Operating profit margin", 'Book to market ratio', 'Short financial leverage', 'Long financial leverage', 'isSmall', 'isMedium']]

In [655]:
df_regL

firm       q  Stock annual return       F1r      EPSP  \
90       TXG  2000q1             0.000000  0.000000  0.000000   
91       TXG  2000q2             0.000000  0.000000  0.000000   
92       TXG  2000q3             0.000000  0.000000  0.000000   
93       TXG  2000q4             0.000000  0.000000  0.000000   
94       TXG  2001q1             0.000000  0.000000  0.000000   
...      ...     ...                  ...       ...       ...   
323721  ZYNE  2021q2             0.450906  0.247615 -0.045058   
323722  ZYNE  2021q3             0.247615 -0.136132 -0.058349   
323723  ZYNE  2021q4            -0.136132 -0.819027 -0.074402   
323724  ZYNE  2022q1            -0.819027 -1.534790 -0.101221   
323725  ZYNE  2022q2            -1.534790  0.000000 -0.182640   

        Sales anual growht  Operating profit growht  Operating profit margin  \
90                     0.0                 0.000000                      0.0   
91                     0.0                 0.000000                      0.0   
92                     0.0                 0.000000                      0.0   
93                     0.0                 0.000000                      0.0   
94                     0.0                 0.000000                      0.0   
...                    ...                      ...                      ...   
323721                 0.0                -0.549558                      NaN   
323722                 0.0                 0.106033                      NaN   
323723                 0.0                -0.132792                      NaN   
323724                 0.0                 0.129250                      NaN   
323725                 0.0                -0.068072                      NaN   

        Book to market ratio  Short financial leverage  \
90                  0.000000                  0.000000   
91                  0.000000                  0.000000   
92                  0.000000                  0.000000   
93                  0.000000                  0.000000   
94                  0.000000                  0.000000   
...                      ...                       ...   
323721              0.401883                  0.002100   
323722              0.449880                  0.002307   
323723              0.597384                  0.002576   
323724              0.748904                  0.002830   
323725              1.151762                  0.003163   

        Long financial leverage  isSmall  isMedium  
90                     0.000000        1         0  
91                     0.000000        1         0  
92                     0.000000        1         0  
93                     0.000000        1         0  
94                     0.000000        1         0  
...                         ...      ...       ...  
323721                 0.004770        1         0  
323722                 0.004569        1         0  
323723                 0.004357        1         0  
323724                 0.003976        1         0  
323725                 0.003543        1         0  

[140841 rows x 13 columns]

In [656]:
df_corrs = df_regL.corr()["Stock annual return"]
df_corrs.sort_values(ascending=False)

Stock annual return         1.000000
F1r                         0.695838
isMedium                    0.100521
Operating profit margin     0.005460
Operating profit growht     0.003944
Sales anual growht          0.001824
Book to market ratio        0.000019
EPSP                        0.000017
Long financial leverage    -0.000229
Short financial leverage   -0.001970
isSmall                    -0.181240
Name: Stock annual return, dtype: float64

In [657]:
from scipy.stats.mstats import winsorize

# Creación del modelo

In [658]:
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from scipy import stats
import statsmodels.api as sm

In [659]:
df_train = df_regL.copy()

In [660]:
df_train = df_train.replace([np.nan], 0) 

In [661]:
df_train.replace([np.inf, -np.inf], np.nan, inplace=True) 

Obtención de variables que se utilizarán para el modelo

In [662]:
df_vals = df_train[["EPSP","Operating profit growht","Short financial leverage","Sales anual growht", 'isSmall', 'isMedium', 'q']]

Filtrar a solo a el penultimo cuarter

In [663]:
df_fd=df_vals.loc[(df_vals["q"] != "2022q1")].index
df_vals=df_vals.drop(df_fd)

In [664]:
df_fd=df_train.loc[(df_train["q"] != "2022q1")].index
df_train=df_train.drop(df_fd)

In [665]:
df_vals = df_vals.drop(['q'], axis=1)

Obtención de las betas

In [666]:
df_vals['b'] = 1

In [667]:
first_column = df_vals.pop('b')
  
df_vals.insert(0, 'b', first_column)

In [668]:
t_x = df_vals.transpose()

In [669]:
xt_x = t_x.dot(df_vals)

In [670]:
inv = np.linalg.inv(xt_x)

In [671]:
xt_y = t_x.dot(df_train["F1r"]) 

In [672]:
B = inv.dot(xt_y)

Valores de betas

In [673]:
B

array([-2.35037504e-01,  3.21310414e+00, -2.97322957e-03, -5.04199211e-02,
       -8.73885530e-04, -3.96927711e-01, -1.71784877e-01])

In [674]:
bethas = pd.DataFrame(B)
bethas = bethas.transpose()
bethas.columns = ['b','EPSP','Operating profit growht','Short financial leverage', 'Sales anual growht', 'isSmall', 'isMedium']
bethas.transpose()

0
b                        -0.235038
EPSP                      3.213104
Operating profit growht  -0.002973
Short financial leverage -0.050420
Sales anual growht       -0.000874
isSmall                  -0.396928
isMedium                 -0.171785

In [675]:
Y = df_train["F1r"]

In [676]:
X2 = sm.add_constant(df_vals)
est = sm.OLS(Y,X2)
est2 = est.fit()
print(est2.summary())

                            OLS Regression Results                            
Dep. Variable:                    F1r   R-squared:                       0.396
Model:                            OLS   Adj. R-squared:                  0.394
Method:                 Least Squares   F-statistic:                     170.6
Date:                Sat, 15 Oct 2022   Prob (F-statistic):          7.26e-167
Time:                        00:27:05   Log-Likelihood:                -1382.2
No. Observations:                1565   AIC:                             2778.
Df Residuals:                    1558   BIC:                             2816.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
b                       

/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


COMO SE PUEDE OBSERVAR, LOS RESULTADOS CON ÁLGEBRA MATRICIAL SON LOS MISMOS QUE CON EL USO DE FRAMEWORKS.

SE PUEDE OBSERVAR EN ESTOS CASOS QUE CASI TODAS LAS VARIABLES TIENEN UNA RELACIÓN POSITIVA CON RESPECTO A SU MAGNITUD, LOS CUALES SON: OPERATINFG PROFIT GROWHT, SHORT FINANCIAL LEVERAGE, SALES ANUAL GROWHT, CON VALORES -0.0030, -0.0530, -0.0009, -0.3969, -0.1718. LA UNICA CON UNA MAGNITUD POSITIVA ES EPSP.



RECORDEMOS QUE LA MAGNITUD ES EL CAMBIO DE LA INDEPENDIENTE CUANDO LA DEPENDIENTE CAMBIA EN 1 UNIDAD. POR EJEMPLO, CUANDO EL "EPSP" CAMBIA EN 1 UNIDAD EL 'F1r' CAMBIA EN 1.18 UNIDADES.

EXISTEN TAMBIEN LAS MAGNITUDES DE LOS TAMAÑOS DE LA EMPRESA, PERO COMO ESTA VARIABLE ES BINARIA SOLAMENTE SE LE SUMARÁ AL INTERCEPTO, DANDONOS COMO RESULTADO 3 DISTINTOS MODELOS, UNO PARA CADA TAMAÑO DE EMPRESA.

# Matriz de varianza y covarianza

In [677]:
df_cov_vals = df_train[["EPSP","Operating profit growht","Short financial leverage","Sales anual growht", 'isSmall', 'isMedium', 'F1r']]

In [678]:
df_cov_vals

EPSP  Operating profit growht  Short financial leverage  \
178    -0.004894                 3.085258                  0.007612   
628    -0.021430                 0.575324                  0.005400   
718    -0.040538                24.978416                  0.006757   
988     0.015805                -0.177031                  0.046625   
1078   -0.053940                 0.617950                  0.006269   
...          ...                      ...                       ...   
322554  0.004833                -0.253012                  0.046930   
322644 -0.043494                 0.000000                  0.007905   
322914  0.006752                 0.054960                  0.097403   
323544  0.006535                -0.441476                  0.005006   
323724 -0.101221                 0.129250                  0.002830   

        Sales anual growht  isSmall  isMedium       F1r  
178               0.081978        0         0 -1.464993  
628               0.328974        1         0 -0.776435  
718               0.000000        0         1 -1.550475  
988              -0.002486        0         0 -0.393073  
1078             -0.390500        1         0 -1.238333  
...                    ...      ...       ...       ...  
322554           -0.099708        0         0 -0.384678  
322644            0.000000        0         1  0.000000  
322914            0.061464        0         0 -0.075013  
323544           -0.544573        0         0  0.093173  
323724            0.000000        1         0 -1.534790  

[1565 rows x 7 columns]

In [679]:
xt = df_cov_vals.transpose()

In [680]:
n = len(df_cov_vals)

In [681]:
ones = np.ones((n,1))

In [682]:
xt1 = xt.dot(ones)

In [683]:
xtx = xt.dot(df_cov_vals)

In [684]:
xt1t = xt1.transpose()

In [685]:
VarCov = (1/(n-1))*(xtx - (1/n)*(xt1.dot(xt1t)) )

In [686]:
df_cov_vals.cov()

EPSP  Operating profit growht  \
EPSP                      0.013539                -0.017386   
Operating profit growht  -0.017386               191.432482   
Short financial leverage -0.001592                -0.044949   
Sales anual growht        0.116067                 2.650270   
isSmall                  -0.026455                 0.026046   
isMedium                  0.011529                -0.205699   
F1r                       0.052053                -0.600087   

                          Short financial leverage  Sales anual growht  \
EPSP                                     -0.001592            0.116067   
Operating profit growht                  -0.044949            2.650270   
Short financial leverage                  0.014497           -0.050491   
Sales anual growht                       -0.050491         2454.581464   
isSmall                                   0.003485           -0.088603   
isMedium                                 -0.003064           -0.229245   
F1r                                      -0.006524           -1.702871   

                           isSmall  isMedium       F1r  
EPSP                     -0.026455  0.011529  0.052053  
Operating profit growht   0.026046 -0.205699 -0.600087  
Short financial leverage  0.003485 -0.003064 -0.006524  
Sales anual growht       -0.088603 -0.229245 -1.702871  
isSmall                   0.247243 -0.133175 -0.160438  
isMedium                 -0.133175  0.209492  0.054882  
F1r                      -0.160438  0.054882  0.567820

In [687]:
VarCov

EPSP  Operating profit growht  \
EPSP                      0.013539                -0.017386   
Operating profit growht  -0.017386               191.432482   
Short financial leverage -0.001592                -0.044949   
Sales anual growht        0.116067                 2.650270   
isSmall                  -0.026455                 0.026046   
isMedium                  0.011529                -0.205699   
F1r                       0.052053                -0.600087   

                          Short financial leverage  Sales anual growht  \
EPSP                                     -0.001592            0.116067   
Operating profit growht                  -0.044949            2.650270   
Short financial leverage                  0.014497           -0.050491   
Sales anual growht                       -0.050491         2454.581464   
isSmall                                   0.003485           -0.088603   
isMedium                                 -0.003064           -0.229245   
F1r                                      -0.006524           -1.702871   

                           isSmall  isMedium       F1r  
EPSP                     -0.026455  0.011529  0.052053  
Operating profit growht   0.026046 -0.205699 -0.600087  
Short financial leverage  0.003485 -0.003064 -0.006524  
Sales anual growht       -0.088603 -0.229245 -1.702871  
isSmall                   0.247243 -0.133175 -0.160438  
isMedium                 -0.133175  0.209492  0.054882  
F1r                      -0.160438  0.054882  0.567820

De igual manera los resultados son iguales entre álgebra matricial y el framework.

Esta matriz representa que tanto varia una variable con respecto a la otra, en caso de compararse consigo misma se compara con su promedio.

# Matriz de correlación

OBTENCION DE LA DESVIACION ESRÁNDAR

In [688]:
var = np.array([])
for i in range(len(VarCov)) :
  var = np.append(var, VarCov.iloc[i,i])


In [689]:
vars = pd.DataFrame(var)
vars = vars.transpose()

In [690]:
vars.columns = ['EPSP','Operating profit growht','Short financial leverage', 'Sales anual growht', 'isSmall', 'isMedium', 'Fir']

In [691]:
stds = var ** 0.5

In [692]:
stds

array([ 0.11635786, 13.83591275,  0.1204023 , 49.54373284,  0.49723503,
        0.45770332,  0.75353839])

In [693]:
stdMat = [[ 0 for y in range( len(stds) ) ] for x in range( len(stds) )]
for i in range(len(stds)):
  for j in range(len(stds)):
    stdMat[i][j] = stds[i] * stds[j]


In [694]:
cov = VarCov.to_numpy()

OBTENCION DE LA MATRIZ DE CORRELACION

In [695]:
corrMat = cov / stdMat
corrMat = pd.DataFrame(corrMat)

In [696]:
corrMat.columns = ['EPSP','Operating profit growht','Short financial leverage', 'Sales anual growht', 'isSmall', 'isMedium', 'Fir']

In [697]:
corrMat = corrMat.transpose()
corrMat.columns = ['EPSP','Operating profit growht','Short financial leverage', 'Sales anual growht', 'isSmall', 'isMedium', 'Fir']

In [698]:
corrMat

EPSP  Operating profit growht  \
EPSP                      1.000000                -0.010799   
Operating profit growht  -0.010799                 1.000000   
Short financial leverage -0.113609                -0.026982   
Sales anual growht        0.020134                 0.003866   
isSmall                  -0.457242                 0.003786   
isMedium                  0.216469                -0.032482   
Fir                       0.593674                -0.057557   

                          Short financial leverage  Sales anual growht  \
EPSP                                     -0.113609            0.020134   
Operating profit growht                  -0.026982            0.003866   
Short financial leverage                  1.000000           -0.008464   
Sales anual growht                       -0.008464            1.000000   
isSmall                                   0.058214           -0.003597   
isMedium                                 -0.055600           -0.010109   
Fir                                      -0.071910           -0.045613   

                           isSmall  isMedium       Fir  
EPSP                     -0.457242  0.216469  0.593674  
Operating profit growht   0.003786 -0.032482 -0.057557  
Short financial leverage  0.058214 -0.055600 -0.071910  
Sales anual growht       -0.003597 -0.010109 -0.045613  
isSmall                   1.000000 -0.585161 -0.428193  
isMedium                 -0.585161  1.000000  0.159126  
Fir                      -0.428193  0.159126  1.000000

In [699]:
df_cov_vals.corr()

EPSP  Operating profit growht  \
EPSP                      1.000000                -0.010799   
Operating profit growht  -0.010799                 1.000000   
Short financial leverage -0.113609                -0.026982   
Sales anual growht        0.020134                 0.003866   
isSmall                  -0.457242                 0.003786   
isMedium                  0.216469                -0.032482   
F1r                       0.593674                -0.057557   

                          Short financial leverage  Sales anual growht  \
EPSP                                     -0.113609            0.020134   
Operating profit growht                  -0.026982            0.003866   
Short financial leverage                  1.000000           -0.008464   
Sales anual growht                       -0.008464            1.000000   
isSmall                                   0.058214           -0.003597   
isMedium                                 -0.055600           -0.010109   
F1r                                      -0.071910           -0.045613   

                           isSmall  isMedium       F1r  
EPSP                     -0.457242  0.216469  0.593674  
Operating profit growht   0.003786 -0.032482 -0.057557  
Short financial leverage  0.058214 -0.055600 -0.071910  
Sales anual growht       -0.003597 -0.010109 -0.045613  
isSmall                   1.000000 -0.585161 -0.428193  
isMedium                 -0.585161  1.000000  0.159126  
F1r                      -0.428193  0.159126  1.000000

LOS RESULTADOS SON LOS MISMOS ENTRE AMBAS SALIDAS.
RECORDEMOS QUE LA CORRELACIÓN ES QUE TANTO SE RELACIONAN LAS VARIALES ENTRE CADA UNA DE ELLAS.

# MULTICOLINEALIDAD

In [700]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

vif_data = pd.DataFrame()
vif_data["feature"] = df_cov_vals.columns
  
# calculating VIF for each feature
vif_data["VIF"] = [variance_inflation_factor(df_cov_vals.values, i)
                          for i in range(len(df_cov_vals.columns))]

print(vif_data)

                    feature       VIF
0                      EPSP  1.936446
1   Operating profit growht  1.007234
2  Short financial leverage  1.065732
3        Sales anual growht  1.008243
4                   isSmall  2.009876
5                  isMedium  1.148296
6                       F1r  2.679952


SE PUEDE OBSERVAR QUE ENTRE NUNGANA DE LAS VARIABLES EXISTE MULTICOLINEALIDAD, ESTO SE DEFINE PORQUE NINGUNA DE ELLAS ES MAYOR A 10

In [701]:
%%shell
jupyter nbconvert --to html /content/gdrive/MyDrive/IA/Estadistica/Entregable2.ipynb

[NbConvertApp] Converting notebook /content/gdrive/MyDrive/IA/Estadistica/Entregable2.ipynb to html
[NbConvertApp] Writing 416379 bytes to /content/gdrive/MyDrive/IA/Estadistica/Entregable2.html
